pykafka/managedbalancedconsumer.py

```diff
@@ -0,0 +1,368 @@
+        :type post_rebalance_callback: function
+        :param use_rdkafka: Use librdkafka-backed consumer if available
+        :type use_rdkafka: bool
+        :param compacted_topic: Set to read from a compacted topic. Forces
+            consumer to use less stringent message ordering logic because compacted
+            topics do not provide offsets in stict incrementing order.
+        :type compacted_topic: bool
+        :param heartbeat_interval_ms: The amount of time in milliseconds to wait between
+            heartbeat requests
+        :type heartbeat_interval_ms: int
+        """
+
+        self._cluster = cluster
+        if not isinstance(consumer_group, bytes):
+            raise TypeError("consumer_group must be a bytes object")
+        self._consumer_group = consumer_group
+        self._topic = topic
+
+        self._auto_commit_enable = auto_commit_enable
+        self._auto_commit_interval_ms = auto_commit_interval_ms
+        self._fetch_message_max_bytes = fetch_message_max_bytes
+        self._fetch_min_bytes = fetch_min_bytes
+        self._num_consumer_fetchers = num_consumer_fetchers
+        self._queued_max_messages = queued_max_messages
+        self._fetch_wait_max_ms = fetch_wait_max_ms
+        self._consumer_timeout_ms = consumer_timeout_ms
+        self._offsets_channel_backoff_ms = offsets_channel_backoff_ms
+        self._offsets_commit_max_retries = offsets_commit_max_retries
+        self._auto_offset_reset = auto_offset_reset
+        self._reset_offset_on_start = reset_offset_on_start
+        self._rebalance_max_retries = rebalance_max_retries
+        self._rebalance_backoff_ms = rebalance_backoff_ms
+        self._post_rebalance_callback = post_rebalance_callback
+        self._is_compacted_topic = compacted_topic
+        self._heartbeat_interval_ms = heartbeat_interval_ms
+        if use_rdkafka is True:
+            raise ImportError("use_rdkafka is not available for {}".format(
+                self.__class__.__name__))
+        self._use_rdkafka = use_rdkafka
+
+        self._rebalancing_lock = cluster.handler.Lock()
+        self._consumer = None
+        self._consumer_id = b''
+        self._worker_trace_logged = False
+        self._worker_exception = None
+
+        if auto_start is True:
+            self.start()
+
+    def _setup_heartbeat_worker(self):
+        """Start the heartbeat worker"""
+        self = weakref.proxy(self)
+
+        def fetcher():
+            while True:
+                try:
+                    if not self._running:
+                        break
+                    self._send_heartbeat()
+                    self._cluster.handler.sleep(self._heartbeat_interval_ms / 1000)
+                except ReferenceError:
+                    break
+                except Exception:
+                    # surface all exceptions to the main thread
+                    self._worker_exception = sys.exc_info()
+                    break
```

Comment:
Do we care enough that if `_send_heartbeat()` fails, we'd want to abort the consumer altogether? I'm not really sure what my answer would be, but IIUC it's a bit like the decision we made earlier for `BalancedConsumer`: "if you're not certain if you're still a group member, just keep consuming until it's clear you're not"?